pkg/kubelet/kuberuntime/sync_action.go

```diff
@@ -0,0 +1,216 @@
+		}
+		if _, ok := appContainers[s.Name]; ok {
+			appContainerStatuses = append(appContainerStatuses, s)
+		}
+	}
+
+	return &statusGroup{
+		sandboxStatuses:       podStatus.SandboxStatuses,
+		initContainerStatuses: initContainerStatuses,
+		appContainerStatuses:  appContainerStatuses,
+	}
+}
+
+// containerKillInfo describes a running container that needs to be killed, and the reason for that.
+type containerKillInfo struct {
+	// The ID of the container, runtime uses it to kill the container.
+	id kubecontainer.ContainerID
+	// The name of the container, used for logging.
+	name string
+	// The reason why the container will be killed, e.g. due to failures.
+	reason string
+	// The boolean indicates whether the container will be restarted.
+	restart bool
+
+	// The status of the container, could used by the starting phase
+	// to know the attempt count and finish timestamp for back-off.
+	// TODO(yifan): This has some duplicated info including id and name.
+	// Figure out if we want to merge two.
+	status *kubecontainer.ContainerStatus
+}
+
+// containerStartInfo describes a container that needs to be started, and the reason for that.
+type containerStartInfo struct {
+	// The pointer to the API container that needs to be started.
+	container *api.Container
+	// The reason why the container will be started, e.g. due to restart.
+	// Used for logging.
+	reason string
+	// The last status of the container if it exits.
+	// This is used to determine the "attempt count" and whether we need to do backoff.
+	status *kubecontainer.ContainerStatus
+}
+
+// sandboxKillInfo describes a sandbox that needs to be killed, and the reason for that.
+type sandboxKillInfo struct {
+	// The ID of the sandbox, runtime uses it to kill the sandbox.
+	id string
+	// The name of the sandbox, used for logging.
+	name string
+	// The reason why the sandbox will be killed. e.g. due to failures.
+	reason string
+	// The boolean indicates whether the sandbox will be restarted.
+	restart bool
+}
+
+// sandboxStartInfo describes a sandbox that needs to be started, and the reason for that.
+type sandboxStartInfo struct {
+	// The pointer to the API pod that needs to be started.
+	sandbox *api.Pod
+	// The number of the attempts that we have tried so far to start
+	// the sandbox.
+	attempt int
+	// The reason why the sandbox will be started. e.g. due to restart.
+	reason string
+}
+
+// sandboxContext includes the information that are necessary in the container creation phase.
+type sandboxContext struct {
+	// The pod references the API pod object for this sync action.
+	pod *api.Pod
+	// The status of the sandbox.
+	sandboxStatus *runtimeapi.PodSandboxStatus
+	// The pull secrets.
+	pullSecrets []api.Secret
+}
+
+// syncAction describes the actions that needs to be taken for one sync pod iteration.
+type syncAction struct {
+	containersToKill  []*containerKillInfo
+	containersToStart []*containerStartInfo // Take cares of the init containers.
```

Comment:
> if we need to start one init container, then containersToStart will only contain one init container, no other containers.

@yifan-gu Didn't notice there is such assumption. lgtm for processing init containers this way.